keepassxc-browser/background/keepass.js

```diff
@@ -398,39 +240,23 @@ keepass.associate = async function(tab) {
             idKey: idKey
         };
 
-        const request = keepass.buildRequest(kpAction, keepass.encrypt(messageData, nonce), nonce, keepass.clientID, true);
-
-        const response = await keepass.sendNativeMessage(request);
-        if (response.message && response.nonce) {
-            const res = keepass.decrypt(response.message, response.nonce);
-            if (!res) {
-                keepass.handleError(tab, kpErrors.CANNOT_DECRYPT_MESSAGE);
-                return AssociatedAction.NOT_ASSOCIATED;
-            }
-
-            const message = nacl.util.encodeUTF8(res);
-            const parsed = JSON.parse(message);
-            keepass.setcurrentKeePassXCVersion(parsed.version);
-            const id = parsed.id;
-
-            if (!keepass.verifyResponse(parsed, incrementedNonce)) {
-                keepass.handleError(tab, kpErrors.ASSOCIATION_FAILED);
-                return AssociatedAction.NOT_ASSOCIATED;
-            } else {
-                // Use public key as identification key with older KeePassXC releases
-                const savedKey = keepass.compareVersion('2.3.4', keepass.currentKeePassXC) ? idKey : key;
-                keepass.setCryptoKey(id, savedKey); // Save the new identification public key as id key for the database
-                keepass.associated.value = true;
-                keepass.associated.hash = parsed.hash || 0;
-            }
+        const response = await keepassClient.sendMessage(kpAction, messageData, nonce, false, true);
+        const [ res, parsed ] = keepassClient.handleResponse(response, incrementedNonce, tab);
```

Comment:
This pattern is repeated often, perhaps these should be done in one function call. Additionally, the keepassClient can calculate the incrementedNonce in that function call. You would need to pass the current tab to the sendMessage function (maybe).

Reply:
Fixed this, and issues above.